File: Claude/code/replication/replicate_main.py
"""replicate_main.py — INDEPENDENT replication of the four headline ATTs.
Uses pandas, statsmodels, numpy. Run: python3 replicate_main.py
"""
import pandas as pd
import numpy as np
import statsmodels.api as sm
from statsmodels.formula.api import ols, glm

NSW_PATH = "/Users/scunning/mixtape/nsw_mixtape.dta"
CPS_PATH = "/Users/scunning/mixtape/nswcps.dta"

nsw = pd.read_stata(NSW_PATH)
cps = pd.read_stata(CPS_PATH)
cps = cps.rename(columns={"treated": "treat", "hispanic": "hisp", "married": "marr"})

# (1) Experimental ATT
exp_att = nsw.loc[nsw.treat == 1, "re78"].mean() - nsw.loc[nsw.treat == 0, "re78"].mean()

# (2) LaLonde OLS
ols_fit = ols("re78 ~ treat + age + educ + black + hisp + marr + nodegree + re74 + re75",
              data=cps).fit()
lalonde_ols = ols_fit.params["treat"]

# (3) HIT local-linear DiD
# NOTE: The deck's figures use the original quadratic-terms PS spec, but
# Python's statsmodels GLM lands at a different local optimum than R's glm()
# on that near-quasi-separated design. This replication uses a leaner spec
# (well-conditioned across all three languages), per the referee2 audit.
from statsmodels.formula.api import logit
ps_formula = "treat ~ age + educ + black + hisp + marr + nodegree + re74 + re75"
ps_fit = logit(ps_formula, data=cps).fit(method="newton", maxiter=200, disp=0)
cps["ps"] = ps_fit.predict(cps)
ps_min = max(cps.loc[cps.treat == 1, "ps"].min(), 0.01)
ps_max = min(cps.loc[cps.treat == 1, "ps"].max(), 0.99)
cs = cps[(cps.ps >= ps_min) & (cps.ps <= ps_max)].copy()
trt = cs[cs.treat == 1].copy(); trt["dy"] = trt.re78 - trt.re75
ctrl = cs[cs.treat == 0].copy(); ctrl["dy"] = ctrl.re78 - ctrl.re75
sd_p = ctrl.ps.std(); iqr_p = ctrl.ps.quantile(0.75) - ctrl.ps.quantile(0.25)
h = 0.9 * min(sd_p, iqr_p/1.34) * (len(ctrl)**(-1/5))

def local_linear_dy(p0):
    k = np.exp(-0.5 * ((ctrl.ps.values - p0)/h)**2) / (h*np.sqrt(2*np.pi))
    if k.sum() < 1e-10:
        return np.nan
    X = np.column_stack([np.ones(len(ctrl)), ctrl.ps.values - p0])
    W = np.diag(k)
    XtWX = X.T @ W @ X
    if np.linalg.det(XtWX) < 1e-12:
        return float(np.average(ctrl.dy.values, weights=k))
    b = np.linalg.solve(XtWX, X.T @ W @ ctrl.dy.values)
    return float(b[0])

trt["dy0"] = trt.ps.apply(local_linear_dy)
hit_att = (trt.dy - trt.dy0).mean()

# (4) Abadie 2005 PS-weighted DiD
cs["dY"] = cs.re78 - cs.re75
p_treat = cs.treat.mean()
w = (cs.treat - cs.ps) / (1 - cs.ps)
abadie_att = (w * cs.dY).mean() / p_treat

print("===== Replication in Python =====")
print(f"  Experimental ATT:   ${exp_att:>10,.0f}")
print(f"  LaLonde OLS:        ${lalonde_ols:>10,.0f}")
print(f"  HIT local-linear:   ${hit_att:>10,.0f}")
print(f"  Abadie 2005:        ${abadie_att:>10,.0f}")

File: Claude/code/replication/replicate_main.R
## replicate_main.R — INDEPENDENT replication of the four headline ATTs
## from the LaLonde-arc deck. Uses only base R + haven where possible.
## Run: Rscript replicate_main.R

suppressPackageStartupMessages(library(haven))

nsw_path <- "/Users/scunning/mixtape/nsw_mixtape.dta"
cps_path <- "/Users/scunning/mixtape/nswcps.dta"

nsw <- read_dta(nsw_path)
cps <- read_dta(cps_path)
names(cps)[names(cps)=="treated"]  <- "treat"
names(cps)[names(cps)=="hispanic"] <- "hisp"
names(cps)[names(cps)=="married"]  <- "marr"

## (1) Experimental ATT — simple diff of means on NSW
exp_att <- mean(nsw$re78[nsw$treat==1]) - mean(nsw$re78[nsw$treat==0])

## (2) LaLonde OLS — drop NSW controls, paste in CPS, OLS with all covariates
ols_fit <- lm(re78 ~ treat + age + educ + black + hisp + marr + nodegree + re74 + re75,
              data=cps)
lalonde_ols <- coef(ols_fit)["treat"]

## (3) HIT estimator — local-linear DiD using only base R
## NOTE: The deck's figures use the original quadratic-terms PS spec.
## This replication uses a leaner spec that converges identically across
## R / Python / Stata; the cross-language gap on the quadratic spec is itself
## a Smith-Todd specification-dependence finding (see the deck's "neighborhood,
## not a point" slide).
ps_fit <- glm(treat ~ age + educ + black + hisp + marr + nodegree + re74 + re75,
              data=cps, family=binomial)
cps$ps <- predict(ps_fit, type="response")
ps_min <- max(min(cps$ps[cps$treat==1]), 0.01)
ps_max <- min(max(cps$ps[cps$treat==1]), 0.99)
cs <- subset(cps, ps>=ps_min & ps<=ps_max)
trt  <- subset(cs, treat==1); trt$dy  <- trt$re78  - trt$re75
ctrl <- subset(cs, treat==0); ctrl$dy <- ctrl$re78 - ctrl$re75
h <- 0.9 * min(sd(ctrl$ps), IQR(ctrl$ps)/1.34) * length(ctrl$ps)^(-1/5)
ll <- function(p0) {
  k <- dnorm((ctrl$ps - p0)/h)
  X <- cbind(1, ctrl$ps - p0); W <- k
  XtWX <- crossprod(X * sqrt(W))
  XtWy <- crossprod(X * W, ctrl$dy)
  if (det(XtWX) < 1e-12) return(weighted.mean(ctrl$dy, k))
  drop(solve(XtWX, XtWy)[1])
}
trt$dy0 <- sapply(trt$ps, ll)
hit_att <- mean(trt$dy - trt$dy0, na.rm=TRUE)

## (4) Abadie 2005 PS-weighted DiD  (same leaner PS spec, refit on full sample)
ps_fit2 <- glm(treat ~ age + educ + black + hisp + marr + nodegree + re74 + re75,
               data=cps, family=binomial)
cps$ps_full <- predict(ps_fit2, type="response")
cs2 <- subset(cps, ps_full > 0.01 & ps_full < 0.99)
cs2$dY <- cs2$re78 - cs2$re75
n1_2 <- sum(cs2$treat); p_treat2 <- n1_2 / nrow(cs2)
w <- (cs2$treat - cs2$ps_full)/(1 - cs2$ps_full)
abadie_att <- mean(w * cs2$dY) / p_treat2

cat("===== Replication in R =====\n")
cat(sprintf("  Experimental ATT:   $%s\n", format(round(exp_att), big.mark=",")))
cat(sprintf("  LaLonde OLS:        $%s\n", format(round(lalonde_ols), big.mark=",")))
cat(sprintf("  HIT local-linear:   $%s\n", format(round(hit_att), big.mark=",")))
cat(sprintf("  Abadie 2005:        $%s\n", format(round(abadie_att), big.mark=",")))
